torchaudio/transforms.py

```diff
@@ -610,6 +610,282 @@ def forward(self, waveform: Tensor) -> Tensor:
+        # zeros per block.
+        zeros_per_block = min(input_sr, output_sr) * cutoff_ratio
+
+        # The convolutional kernel size will be n = (blocks_per_side*2 + 1),
+        # i.e. we add that many blocks on each side of the central block.  The
+        # window radius (defined as distance from center to edge)
+        # is `blocks_per_side` blocks.  This ensures that each sample in the
+        # central block can "see" all the samples in its window.
+        #
+        # Assuming the following division is not exact, adding 1
+        # will have the same effect as rounding up.
+        blocks_per_side = 1 + int(num_zeros / zeros_per_block)
+
+        kernel_width = 2 * blocks_per_side + 1
+
+        # Computations involving time will be in units of 1 block.  Actually this
+        # is the same as the `canonical` time axis since each block has input_sr
+        # input samples, so it would be one of whatever time unit we are using
+        window_radius_in_blocks = blocks_per_side
+
+        # The `times` below will end up being the args to the sinc function.
+        # For the shapes of the things below, look at the args to `view`.  The terms
+        # below will get expanded to shape (output_sr, input_sr, kernel_width) through
+        # broadcasting
+        # We want it so that, assuming input_sr == output_sr, along the diagonal of
+        # the central block we have t == 0.
+        # The signs of the output_sr and input_sr terms need to be opposite.  The
+        # sign that the kernel_width term needs to be will depend on whether it's
+        # convolution or correlation, and the logic is tricky.. I will just find
+        # which sign works.
+
+        times = (
+            torch.arange(output_sr, dtype=dtype).reshape((output_sr, 1, 1)) / output_sr -
+            torch.arange(input_sr, dtype=dtype).reshape((1, input_sr, 1)) / input_sr -
+            (torch.arange(kernel_width, dtype=dtype).reshape((1, 1, kernel_width)) - blocks_per_side))
+
+        def hann_window(a):
+            """
+            hann_window returns the Hann window on [-1,1], which is zero
+            if a < -1 or a > 1, and otherwise 0.5 + 0.5 cos(a*pi).
+            This is applied elementwise to a, which should be a NumPy array.
+
+            The heaviside function returns (a > 0 ? 1 : 0).
+            """
+            #return np.heaviside(1 - np.abs(a), 0.0) * (0.5 + 0.5 * np.cos(a * np.pi))
+            pi = torch.acos(torch.zeros(1)).item() * 2
+            return torch.heaviside(1 - torch.abs(a), 0.0) * (0.5 + 0.5 * torch.cos(a * pi))
+
+
+        def kaiser_window(a, beta):
+            #TODO: the kaiser window can be added once scipy.special.i0 has been implement in torch
+            #w = scipy.special.i0(beta * np.sqrt(np.clip(1 - ((a - 0.0) / 1.0) ** 2.0, 0.0, 1.0))) / special.i0(beta)
+            #return np.heaviside(1 - np.abs(a), 0.0) * w
+            return None
+
+
+        def sinc(x):
```

Comment:
Looks like `sinc` function is under development.

Reply:
Oh good. I'll leave the definition of the sinc function as it is for now, but the new function should be used when it's available.